agent/stats/engine.go

```diff
@@ -1077,3 +1079,72 @@ func (engine *DockerStatsEngine) SetPublishServiceConnectTickerInterval(publishS
 func (engine *DockerStatsEngine) GetPublishMetricsTicker() *time.Ticker {
 	return engine.publishMetricsTicker
 }
+
+func (engine *DockerStatsEngine) getEBSVolumeMetrics(taskArn string) []*ecstcs.VolumeMetric {
+	task, err := engine.resolver.ResolveTaskByARN(taskArn)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to get corresponding task from dd with task arn: %s", taskArn))
+		return nil
+	}
+
+	if !task.IsEBSTaskAttachEnabled() {
+		logger.Debug("Task not EBS-backed, skip gathering EBS volume metrics.", logger.Fields{
+			"taskArn": taskArn,
+		})
+		return nil
+	}
+
+	// TODO: Remove the CSI client from the stats engine and just always have the CSI client created
+	// since a new connection is created regardless and it'll make the stats engine less stateful
+	if engine.csiClient == nil {
+		client := csiclient.NewCSIClient(filepath.Join(csiclient.DefaultSocketHostPath, csiclient.DefaultImageName, csiclient.DefaultSocketName))
+		engine.csiClient = &client
+	}
+	return engine.fetchEBSVolumeMetrics(task, taskArn)
+}
+
+func (engine *DockerStatsEngine) fetchEBSVolumeMetrics(task *apitask.Task, taskArn string) []*ecstcs.VolumeMetric {
+	var metrics []*ecstcs.VolumeMetric
+	for _, tv := range task.Volumes {
+		if tv.Volume.GetType() == taskresourcevolume.EBSVolumeType {
+			volumeId := tv.Volume.GetVolumeId()
+			hostPath := tv.Volume.Source()
+			volumeName := tv.Volume.GetVolumeName()
+			metric, err := engine.getVolumeMetricsWithTimeout(volumeId, hostPath)
+			if err != nil {
+				logger.Error("Failed to gather metrics for EBS volume", logger.Fields{
+					"VolumeId":             volumeId,
+					"SourceVolumeHostPath": hostPath,
+					"Error":                err,
+				})
+				continue
+			}
+			usedBytes := aws.Float64((float64)(metric.Used))
+			totalBytes := aws.Float64((float64)(metric.Capacity))
+			metrics = append(metrics, &ecstcs.VolumeMetric{
+				VolumeId:   aws.String(volumeId),
+				VolumeName: aws.String(volumeName),
+				Utilized: &ecstcs.UDoubleCWStatsSet{
+					Max:         usedBytes,
+					Min:         usedBytes,
+					SampleCount: aws.Int64(1),
+					Sum:         usedBytes,
+				},
+				Size: &ecstcs.UDoubleCWStatsSet{
+					Max:         totalBytes,
+					Min:         totalBytes,
+					SampleCount: aws.Int64(1),
+					Sum:         totalBytes,
+				},
+			})
+		}
+	}
+	return metrics
+}
+
+func (engine *DockerStatsEngine) getVolumeMetricsWithTimeout(volumeId, hostPath string) (*csiclient.Metrics, error) {
+	derivedCtx, cancel := context.WithTimeout(engine.ctx, time.Second*1)
+	// releases resources if GetVolumeMetrics finishes before timeout
+	defer cancel()
+	return engine.csiClient.GetVolumeMetrics(derivedCtx, volumeId, hostPath)
```

Comment:
Is this method implemented for Windows?

Reply:
Yes the code path is common for both Linux and Windows- https://github.com/aws/amazon-ecs-agent/blob/437e051d75d96f9c5619f8b8742bd84d5b959e9c/ecs-agent/csiclient/csi_client.go#L152

`GetVolumeMetrics` depends upon few methods-
- `NodeGetVolumeStats` from `CSI Driver` which is dependent upon the following-
    - `PathExists` in `mounter` package
    - `IsBlockDevice` in `utils` package
    - `getBlockSizeBytes` from `CSI Driver`
    - `GetMetrics` from `CSI Driver`
        - `getFsInfo`
            - `Info` method from `fs` package

From the list above, the following methods are implemented as part of the PR you raised- https://github.com/aws/amazon-ecs-agent/pull/4214
- `PathExists`
- `getBlockSizeBytes`
- `Info`

`IsBlockDevice` method is part of https://github.com/aws/amazon-ecs-agent/pull/4224.

Reply:
Basically, the idea here is to move what we have in `engine_unix.go` into `engine.go` so that both platforms can use the same implementation (please correct me if this is incorrect). 

>Refactor the code from Linux specific files to common platform files.

Are there any notable changes/refactors that has been made to the existing implementation of `getEBSVolumeMetrics`, `fetchEBSVolumeMetrics`, and `getVolumeMetricsWithTimeout` ? Otherwise, calling it a refactor is a bit misleading IMO.

Reply:
You are correct. This PR just moves the methods from Linux specific files to platform agnostic ones. 
The refactor is from the perspective of moving things from platform to general code files.

If you feel strongly, I can change the commit message and PR title.

Reply:
Not a big blocker for me. Was more so of getting clarification.